src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

```diff
@@ -927,7 +928,7 @@ private class DeclarationAnalysisData
         internal AnalyzerDriver(ImmutableArray<DiagnosticAnalyzer> analyzers, Func<SyntaxNode, TLanguageKindEnum> getKind, AnalyzerManager analyzerManager) : base(analyzers, analyzerManager)
         {
             _getKind = getKind;
-            _declarationAnalysisDataCache = new ConditionalWeakTable<SyntaxReference, DeclarationAnalysisData>();
+            _declarationAnalysisDataCache = new ConditionalWeakTable<Compilation, Dictionary<Location, DeclarationAnalysisData>>();
```

Comment:
@mavasani There are no locks (not even multi-reader locks) in `ConcurrentDictionary<TKey, TValue>.TryGetValue`, which means our fast path can be written without any locks.